### ISSUE TITLE
Creating networks via docker

### ISSUE THREAD

**Issue by squaremo** [docker/libnetwork#150]
There's not much point in having the whole driver subsystem unless one can actually create a network. I don't see that as part of docker/docker#13060.

If it is a requirement to not add new commands to docker (e.g.,`docker network create`), then perhaps it can be part of the syntax of the `--net` argument. For example,

    docker run --net=weave:mynet

where `weave` refers to the driver, and `mynet` is a network created with that driver (if it doesn't already exist).

**Comment by mrjana** [comment]
@squaremo docker/docker#13060 is only the first of a set of PR which will made in there. So the short answer is it's coming. The idea is to implement the majority of the CLI handles and remote API handlers in libnetwork itself and just hook it up to docker core with a small function. Introducing a new UI or remote API in docker needs more discussion and that's why the goal of docker/docker#13060 has been limited to modularize networking code out of docker and providing a clean interface to networking code for docker or anybody else to use.

BTW, for the initial implementation of CLI handlers take a look at `libnetwork/client` code

**Comment by mavenugo** [comment]
Also, we are currently integrating the `libnetwork/client` with `libnetwork/api` as we speak and we will have a `dnet` tool that makes use of these. As @mrjana suggested, the docker CLI integration will follow soon after #13060 is merged.

**Comment by squaremo** [comment]
Another requirement: the ability to add more than one network in the same invocation is quite important. Typically one interface will provide internet-in-general access, and one will be the "cluster" network (e.g., weave).

**Comment by dave-tucker** [comment]
@squaremo so assuming weave1 is a  network created using the weave driver, you'd want `--net=bridge --net=weave1`?

**Comment by squaremo** [comment]
> so assuming weave1 is a network created using the weave driver, you'd want --net=bridge --net=weave1?

Yes exactly. One wrinkle is that typically we'd want the bridge driver to set the default gateway, but the weave driver to provide (or at least influence) the `/etc/resolv.conf`.

**Comment by dave-tucker** [comment]
Thanks @squaremo, @mavenugo this seems like a sensible requirement. Are multiple occurrences of the `--net` flag going to be supported in the integration PR mentioned above? If not, can we add them?

**Comment by mavenugo** [comment]
@dave-tucker @squaremo no. I will be pushing https://github.com/mavenugo/docker/commit/c0c7f372375b3bd112c6e0aaa28001f495273782 shortly. This is based on the discussions that i followed for volumes, the idea is to introduce --network-driver instead of overloading the --net string. We need to find a way with these constraints.

**Comment by squaremo** [comment]
@mavenugo Does `--network-driver` coexist with `--net`; i.e., could one use

    docker run --net=bridge --network-driver=weave -ti ubuntu

and expect a bridge interface as well as a weave interface?

**Comment by dave-tucker** [comment]
@mavenugo right, I'm not talking about overloading of `--net` here, I'm talking about supporting multiple occurrences of `--net` assuming that a network has already been created using the `docker network create` CLI

**Comment by shettyg** [comment]
If there are going to be multiple --net's, is there an idea on how to pass network labels per --net?

**Comment by mavenugo** [comment]
@dave-tucker @squaremo @shettyg All are valid and reasonable questions. There are a few trade-offs to be made between simple, consistent UI & functionality. I will get back to you all later today.

**Comment by dave-tucker** [comment]
It hasn't been discussed here yet, but my expectation is that labels are namespaced per [the docs](http://docs.docker.com/userguide/labels-custom-metadata/)

`docker run -it --net=bridge --net=weave1 -net=vmware1 -l works.weave.foo=bar -l com.vmware.baz=quux debian:jessie`

A driver should get passed all labels, but only respond to those in it's namespace

**Comment by squaremo** [comment]
> It hasn't been discussed here yet, but my expectation is that labels are namespaced per the docs
> 
> docker run -it --net=bridge --net=weave1 -net=vmware1 -l works.weave.foo=bar -l com.vmware.baz=quux debian:jessie
> 
> A driver should get passed all labels, but only respond to those in it's namespace

Ah, but the label namespaces correspond to the driver name, rather than the network name. So if a container has two endpoints provided by the same driver, how will that driver know which labels apply to which endpoint?

**Comment by dave-tucker** [comment]
Prefix the label with the network name?  E.g `works.weave.weave1.foo`

**Comment by mrjana** [comment]
@squaremo @dave-tucker @mavenugo  Is it very important to support to support multiple networks in docker run command. `docker run` can be used to join the initial network. There are always going to be `network service join` commands to join additional endpoints after the initial join. I know this is racy to some extent if the application wants to communicate in the "cluster" network immediately but then we are probably risking introducing something hastily here. Please let me know what you guys think?

Also the labels in the `docker run` command are always going to be considered as container labels and are given to the driver only on joins

**Comment by shettyg** [comment]
>>There are always going to be network service join commands to join additional endpoints after the initial join. I know this is racy to some extent if the application wants to communicate in the "cluster" network immediately but then we are probably risking introducing something hastily here.

@mrjana, the raciness may be important IMO. Many badly written applications will simply fail when they can't reach out to peers. So you are effectively expecting applications to be written in a way that IP reachability to a peer should always have retry logic.

**Comment by mrjana** [comment]
@shettyg I am not saying solving the raciness is not important. But it is much more important to get the UI right, otherwise it is very difficult to revert it back.

**Comment by bboreham** [comment]
@shettyg I think the race condition manifests in ways that are worse than "simply fail".  E.g. if your container has a service that listens on 0.0.0.0, it will listen on all interfaces that are active /at the time you do the listen/; it will not add interfaces that are added later.

**Comment by squaremo** [comment]
> I know this is racy to some extent if the application wants to communicate in the "cluster" network immediately

It's already the case that you can add a weave interface to a container after the fact; avoiding the extra command (and the entailed race) is the primary motivation for developing a plugin.

Weave needs containers to have an interface on the bridge network _as_well_, and for this not to be racy or require extra commands, since it's used to provide name resolution.

**Comment by shettyg** [comment]
Since endpoint creation has been separated out, would something like the following not work for anyone?

* docker network create NETWORK --driver=driver1 --labels foo=bar
* docker endpoint create NETWORK --labels alice=bob

With the UUID returned by the endpoint creation (Similar syntax as --net=container:containerid):

* docker run --net=bridge --net=driver1:EP_ID1 --labels driver1.ep_id1=foo --net=driver2:EP_ID2 --labels driver2.ep=bar command

The above only calls join() and skips createendpoint().

**Comment by squaremo** [comment]
@shettyg There is an important difference between an endpoint that is created during `docker run` and an endpoint created _ad-hoc_; the former is garbage collected when the container stops, but the latter will hang around until deleted explicitly. That latter behaviour requires much more diligence on the part of the user, especially since it won't necessarily be obvious when those endpoints can reasonably be deleted.

**Comment by shettyg** [comment]
@squaremo 
I see what you mean. I wonder whether we can do both. For e.g., one could do:
docker run --net=bridge --net=driver1:mynet

In the above case, you would call createendpoint() and join()

Alternatively,
docker run --net=bridge --endpoint=driver1:uuid

In the above case, you would only call join()
My suggestion likely won't fly because now there is a new docker cli option '--endpoint'.

(According to @mrjana, labels provided to createndpoint() and join() are different, so that is another constraint with which this has to work)

**Comment by squaremo** [comment]
I believe https://github.com/docker/docker/pull/13441 will address this issue, in part, but seems caught up in bikeshedding :(

**Comment by shettyg** [comment]
It looks like Solomon's 'docker run --publish-service db.work-dev' is similar to my 'docker run  --endpoint=driver1:uuid'

**Comment by tomdee** [comment]
@mavenugo @dave-tucker can this issue be closed now that it is tracked under https://github.com/docker/docker/issues/14593

**Comment by dave-tucker** [comment]
Thanks @tomdee, closing in favour of docker/docker#14593
